src/components/Postiterie.vue

```diff
@@ -1,203 +1,454 @@
+        canvas.height = imagedata.height;
+        ctx.putImageData(imagedata, 0, 0);
+        let image = new Image();
+        image.src = canvas.toDataURL();
+        return image;
+    },
+
+    /*
+      FUNCTION crop
+      Crop an image using the coordinates and return its imageData
+      IN : x, y (numbers) coordinates from which to start croping
+           width, height (numbers) length to crop
+      OUT: imageData
+    */
+    crop: function( x, y, width, height){
+      let canvas = document.createElement('canvas');
+      let context = canvas.getContext('2d');
+      let img = document.getElementById('preview');
+      canvas.width = img.width;
+      canvas.height = img.height;
+      context.drawImage(img, 0, 0);
+      return context.getImageData(x, y, width, height);
+    },
+
     /*
-      FUNCTION previewImage
-      preview the uploaded image (and activate the button to track)
+      FUNCTION getImage
+      get the uploaded image and start the tracking
       IN : fieldName : the fieldName ;)
            file : the image
     */
-    previewImage: function(fieldName, file) {
-            this.isNoImage = false;
-            let imageFile = file[0] 
-            let formData = new FormData()
-            let imageURL = URL.createObjectURL(imageFile)
-            formData.append(fieldName, imageFile)
+    getImage: function(fieldName, file) {
+            this.link = null;
+            let imageFile = file[0];
+            let formData = new FormData();
+            let imageURL = URL.createObjectURL(imageFile);
+            formData.append(fieldName, imageFile);
             this.image=imageURL;
     },
 
     /*
       FUNCTION tracking
-      Track post-its and draw a square around them
+      Track post-its and add them to results[]
+      Then, send the file
+      TODO : investigate odd behavior --> the function doesn't work if not started from a onclick event (likely due to tracker.on('track', function(event))) )
     */
     tracking: function () {
       
       let $this = this;
-      let img = document.getElementById('preview');
-      let demoContainer = document.querySelector('.container');
+      $this.trackingFinished = false;
+      $this.trackingFinishedWithNoResult = false;
+
       let tracker = new tracking.ColorTracker(['magenta', 'cyan', 'yellow']);
 
       tracker.on('track', function(event) {
+        
         event.data.forEach(function(rect) {
-          window.plot(rect.x, rect.y, rect.width, rect.height, rect.color);
-          let croppedImageData = crop(rect.x,rect.y,rect.width,rect.height);
-          let croppedImage = imageDataToImage(croppedImageData);
+          
+          let croppedImageData = $this.crop(rect.x,rect.y,rect.width,rect.height);
+          let croppedImage = $this.imageDataToImage(croppedImageData);
           $this.results.push(croppedImage);
+          console.log("+1");
+
         });
-      });
 
-      this.isNoResult = false;
+        setTimeout(function(){
```

Comment:
Is the setTimeout necessary? Usually, you better use Promises.

Reply:
Couldn't figure out how to use promises here, I tried using promise.all, but there was a race condition because it always triggered before the foreach. A if inside the foreach resolved the issue, it seems a bit less dirty than a setTimeout

Reply:
`Promise.all` works only if it is given an array of promises.
`forEach` doesn't return anything, you should better use a `map` and return a promise inside, it will result in an array of promises.

Example:
```
const croppedImagePromises = images.map((image)=>{
    const croppingPromise = image.crop(); // here the async call, returning a promise
    // do other things
    return croppingPromise;
});

Promise.all(croppedImagePromises).then(()=>{
    // here we're sure all images are cropped
});
```

If the example doesn't help, let's schedule a call.